drivers: sensor: bma4xx: Implement streaming APIs

cpeggy-25 opened the pull request:
Add a streaming implementation for the BMA4XX using both FIFO watermark and FIFO full interrupts.
A batch duration of 3,000 have been used for verification.
```
uart:~$ sensor attr_set bma4xx@19 all batch_dur 3000
bma4xx@19 channel=all, attr=batch_dur set to value=3000
[00:00:03.111,602] <inf> bma4xx: FIFO ENABLED
uart:~$ sensor stream bma4xx@19 on fifo_wm incl
Enabling stream...
channel type=0(accel_x) index=0 shift=6 num_samples=11 value=12784741210ns (1.398213)
channel type=1(accel_y) index=0 shift=6 num_samples=11 value=12784741210ns (1.398213)
channel type=2(accel_z) index=0 shift=6 num_samples=11 value=12784741210ns (1.398213)
channel type=3(accel_xyz) index=0 shift=6 num_samples=11 value=12784741210ns, (1.398213, 0.315163, 9.528051)
channel type=0(accel_x) index=0 shift=6 num_samples=10 value=12879747314ns (1.403959)
channel type=1(accel_y) index=0 shift=6 num_samples=10 value=12879747314ns (1.403959)
channel type=2(accel_z) index=0 shift=6 num_samples=10 value=12879747314ns (1.403959)
channel type=3(accel_xyz) index=0 shift=6 num_samples=10 value=12879747314ns, (1.403959, 0.256658, 9.482953)
channel type=0(accel_x) index=0 shift=6 num_samples=10 value=12979814453ns (1.415452)
channel type=1(accel_y) index=0 shift=6 num_samples=10 value=12979814453ns (1.415452)
channel type=2(accel_z) index=0 shift=6 num_samples=10 value=12979814453ns (1.415452)
channel type=3(accel_xyz) index=0 shift=6 num_samples=10 value=12979814453ns, (1.415452, 0.342849, 9.525091)
…
…
…
```

teburd reviewed (commented):
Looks good to me, the board overlay for the shell is a bit off though. I don't think we have a great way of dealing with this sort of thing unfortunately. Maybe @MaureenHelm will have some other thoughts, happy to approve once the shell board overlay has been discussed as the RTIO and sensor related changes look good to me.

yperess commented on drivers/sensor/bosch/bma4xx/bma4xx.h:
```diff
@@ -197,45 +75,81 @@ struct bma4xx_hw_operations {
-	uint8_t accel_bwp;
-	/** Current output data rate as a register value */
-	uint8_t accel_odr;
+	struct bma4xx_config cfg;
```
I think there's some confusion around what the `bma4xx_config` is. We're duplicating the bus information.

When you create a driver via `SENSOR_DEVICE_DT_INST_DEFINE` you get to pass 2 pointers: data and config. Data is runtime mutable and config is meant to keep build time per instance information that you'll need later at runtime. In this case the bus is static and should be in the config, while the actual runtime configuration belong in the data.

I think what you want to do is create a new struct `struct bma4xx_runtime_config` that holds all the values of `accel_fs_range`, `accel_bwp`, etc and have that be a part of `bma4xx_data` and use it in the configure and safe configure functions

PeggyCienet replied:
Done, thanks!

yperess reviewed (changes requested):
Can you include a sample output in the PR description?

MaureenHelm reviewed (changes requested):
Each commit should correspond to a self-contained, meaningful change. For example, adding a feature, fixing a bug, or refactoring existing code should be separate commits. Avoid mixing different types of changes (e.g., feature implementation and unrelated refactoring) in the same commit.
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-requirements

PeggyCienet commented on drivers/sensor/bosch/bma4xx/CMakeLists.txt:
```diff
@@ -2,6 +2,15 @@
+    bma4xx.c
+    bma4xx_common.c
+    bma4xx_i2c.c
+    bma4xx_spi.c
```
Done, thanks!

PeggyCienet commented on drivers/sensor/bosch/bma4xx/bma4xx_interrupt.h:
```diff
@@ -0,0 +1,34 @@
+#include <zephyr/device.h>
+#include <zephyr/drivers/sensor.h>
+
+#ifdef CONFIG_BMA4XX_STREAM
```
Done, thanks!

PeggyCienet commented on drivers/sensor/bosch/bma4xx/bma4xx.c:
```diff
@@ -235,456 +260,36 @@ static int bma4xx_chip_init(const struct device *dev)
 	}
-
-	k_sleep(K_USEC(1000));
+#endif
```
Done, thanks!

PeggyCienet commented on drivers/sensor/bosch/bma4xx/bma4xx_spi.c:
```diff
@@ -16,41 +16,39 @@
-	.write_reg  = bma4xx_spi_write_reg,
+	.read_reg = bma4xx_spi_read_reg,
+	.write_reg = bma4xx_spi_write_reg,
 	.update_reg = bma4xx_spi_update_reg,
```
Reverted the changes. Thanks!

PeggyCienet commented on drivers/sensor/bosch/bma4xx/Kconfig:
```diff
@@ -2,32 +2,44 @@
+	depends on EMUL
+	help
+	  Enable the hardware emulator for the BMA4XX. Doing so allows exercising
+	  sensor APIs for this sensor in native_sim and qemu.
```
Removed the unrelated parts/files. Thanks!

yperess reviewed (commented):
This is looking good, but please address Maureen's and Tom's comments.

For Maureen's it'll take a little bit of reworking the commits, but I can think of a few to help you get started:
1. Split off the bma4xx_defs.h
2. Update the configuration to use the safe_config
3. Implement the streaming API

I think also you might not be able to merge the sample (I'll let @MaureenHelm weigh in on this) but since it's not something that's a part of the board and requires people to put the board together I'm not sure what our policies are around that.

yperess commented on drivers/sensor/bosch/bma4xx/bma4xx.h:
```diff
@@ -67,35 +82,110 @@ struct bma4xx_data {
 
+static inline void bma4xx_accel_reg_to_hz(uint8_t odr, struct sensor_value *out)
+{
+	switch (odr) {
```
These ODRs are sequential, instead of adding all these branches you could just create a static array of sensor_value.

```c
static const sensor_value odr_values[] = {
    {0, 0},
    {0, 781250},
    ...
};

if (odr >= ARRAY_SIZE(odr_values)) {
  return -EINVAL;
}
*out = odr_values[odr];
return 0;
```

PeggyCienet replied:
Done, thanks!

yperess reviewed (commented):
The last commit is still a bit large, it seems like we could split it up to
1. moving the common code and decoder to the separate files. This would be an easy change to review since it should be all cut/paste
2. implementing the streaming logic

This will allow reviewers to focus on the new logic during the review. Overall this is looking really good.

yperess reviewed (changes requested):
Previous comment should have been a request for changes, sorry.

PeggyCienet commented on drivers/sensor/bosch/bma4xx/bma4xx_rtio_stream.c:
```diff
@@ -0,0 +1,358 @@
+
+			rtio_sqe_prep_tiny_write(write_signal_path_reset, iodev, RTIO_PRIO_NORM,
+						 write_buffer, ARRAY_SIZE(write_buffer), NULL);
+			rtio_submit(r, 1);
```
Done, thanks!

PeggyCienet commented on drivers/sensor/bosch/bma4xx/bma4xx_rtio.c:
```diff
@@ -0,0 +1,154 @@
+
+	edata->header.timestamp = sensor_clock_cycles_to_ns(cycles);
+
+	rc = bma4xx_get_accel_data(dev, &edata->accel_xyz);
```
Done. Please have a check, thanks!

PeggyCienet commented on drivers/sensor/bosch/bma4xx/bma4xx.c:
```diff
@@ -274,8 +275,8 @@ static int bma4xx_chip_init(const struct device *dev)
 	/* Enable accelerometer */
-	status = bma4xx->hw_ops->update_reg(dev, BMA4XX_REG_POWER_CTRL, BMA4XX_BIT_ACC_EN,
-					    BMA4XX_BIT_ACC_EN);
+	status = bma4xx->hw_ops->update_reg(dev, BMA4XX_REG_POWER_CTRL, BMA4XX_FIFO_ACC_EN,
```
Done, thanks!

PeggyCienet commented on drivers/sensor/bosch/bma4xx/bma4xx_common.c:
```diff
@@ -0,0 +1,102 @@
+	__ASSERT(res == 0, "%s could not enable non-latch mode", __func__);
+
+	/* Enable accelerometer */
+	res |= dev_data->hw_ops->write_reg(dev, BMA4XX_REG_POWER_CTRL,
```
Done, thanks!

Commit:
drivers: sensor: bma4xx: Split off bma4xx_defs.h

Added register definitions, bit masks, and related constants
for the BMA4XX sensor in a separate header file.

Signed-off-by: Peggy Chen <peggy.chen@cienet.com>

Commit:
drivers: sensor: bma4xx: Update config to use bma4xx_safely_configure

Updated the sensor configuration to use the bma4xx_safely_configure
function for safer and more reliable initialization.

Signed-off-by: Peggy Chen <peggy.chen@cienet.com>

Commit:
drivers: sensor: bma4xx: Refactor decoder into separate files

Refactor the sensor decoder code into individual files for
improved maintainability and readability.

Signed-off-by: Peggy Chen <peggy.chen@cienet.com>

Commit:
drivers: sensor: bma4xx: Refactor submit logic into separate files

Refactor the sensor submit logic into individual files
to enhance maintainability and readability.

Signed-off-by: Peggy Chen <peggy.chen@cienet.com>

Commit:
drivers: sensor: bma4xx: Enable and resolve issues with emulator

Enable the sensor emulator and address related issues
to ensure proper functionality.

Signed-off-by: Peggy Chen <peggy.chen@cienet.com>

Commit:
drivers: sensor: bma4xx: Implement streaming APIs

Add a streaming implementation for the BMA4XX using both
FIFO watermark and FIFO full interrupts. A batch duration of
3,000 have been used for verification.

Signed-off-by: Peggy Chen <peggy.chen@cienet.com>

PeggyCienet commented:
Hi @yperess, @MaureenHelm, @teburd,

I have made the necessary updates to the PR based on your feedback and have organized the changes into 6 separate commits. Whenever you have a moment, could you kindly review this PR?

Thank you!
